File: app/src/main/java/com/todoroo/astrid/files/FilesControlSet.java
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */

package com.todoroo.astrid.files;

import static android.app.Activity.RESULT_OK;
import static org.tasks.data.TaskAttachment.createNewAttachment;
import static org.tasks.dialogs.AddAttachmentDialog.REQUEST_AUDIO;
import static org.tasks.dialogs.AddAttachmentDialog.REQUEST_CAMERA;
import static org.tasks.dialogs.AddAttachmentDialog.REQUEST_GALLERY;
import static org.tasks.dialogs.AddAttachmentDialog.REQUEST_STORAGE;
import static org.tasks.dialogs.AddAttachmentDialog.newAddAttachmentDialog;
import static org.tasks.files.FileHelper.copyToUri;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.Nullable;
import butterknife.BindView;
import butterknife.OnClick;
import com.todoroo.astrid.data.Task;
import java.util.ArrayList;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.data.TaskAttachment;
import org.tasks.data.TaskAttachmentDao;
import org.tasks.dialogs.DialogBuilder;
import org.tasks.files.FileHelper;
import org.tasks.injection.ForActivity;
import org.tasks.injection.FragmentComponent;
import org.tasks.preferences.Preferences;
import org.tasks.ui.TaskEditControlFragment;

public class FilesControlSet extends TaskEditControlFragment {

  public static final int TAG = R.string.TEA_ctrl_files_pref;

  private static final String FRAG_TAG_ADD_ATTACHMENT_DIALOG = "frag_tag_add_attachment_dialog";
  private static final char LEFT_TO_RIGHT_MARK = '\u200e';

  @Inject TaskAttachmentDao taskAttachmentDao;
  @Inject DialogBuilder dialogBuilder;
  @Inject @ForActivity Context context;
  @Inject Preferences preferences;

  @BindView(R.id.attachment_container)
  LinearLayout attachmentContainer;

  @BindView(R.id.add_attachment)
  TextView addAttachment;

  private String taskUuid;

  @Nullable
  @Override
  public View onCreateView(
      LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    View view = super.onCreateView(inflater, container, savedInstanceState);

    taskUuid = task.getUuid();

    for (TaskAttachment attachment : taskAttachmentDao.getAttachments(taskUuid)) {
      addAttachment(attachment);
    }

    if (savedInstanceState == null) {
      if (task.hasTransitory(TaskAttachment.KEY)) {
        for (Uri uri : (ArrayList<Uri>) task.getTransitory(TaskAttachment.KEY)) {
          copyToAttachmentDirectory(uri);
        }
      }
    }

    return view;
  }

  @OnClick(R.id.add_attachment)
  void addAttachment(View view) {
    newAddAttachmentDialog(this).show(getFragmentManager(), FRAG_TAG_ADD_ATTACHMENT_DIALOG);
  }

  @Override
  protected int getLayout() {
    return R.layout.control_set_files;
  }

  @Override
  public int getIcon() {
    return R.drawable.ic_outline_attachment_24px;
  }

  @Override
  public int controlId() {
    return TAG;
  }

  @Override
  public void apply(Task task) {}

  @Override
  public void onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == REQUEST_CAMERA || requestCode == REQUEST_AUDIO) {
      if (resultCode == RESULT_OK) {
        Uri uri = data.getData();
        copyToAttachmentDirectory(uri);
        FileHelper.delete(context, uri);
      }
    } else if (requestCode == REQUEST_STORAGE || requestCode == REQUEST_GALLERY) {
      if (resultCode == RESULT_OK) {
        copyToAttachmentDirectory(data.getData());
      }
    } else {
      super.onActivityResult(requestCode, resultCode, data);
    }
  }

  private void addAttachment(TaskAttachment taskAttachment) {
    View fileRow =
        getActivity().getLayoutInflater().inflate(R.layout.file_row, attachmentContainer, false);
    fileRow.setTag(taskAttachment);
    attachmentContainer.addView(fileRow);
    addAttachment(taskAttachment, fileRow);
  }

  private void addAttachment(final TaskAttachment taskAttachment, final View fileRow) {
    TextView nameView = fileRow.findViewById(R.id.file_text);
    String name = LEFT_TO_RIGHT_MARK + taskAttachment.getName();
    nameView.setText(name);
    nameView.setOnClickListener(v -> showFile(taskAttachment));
    View clearFile = fileRow.findViewById(R.id.clear);
    clearFile.setOnClickListener(
        v ->
            dialogBuilder
                .newMessageDialog(R.string.premium_remove_file_confirm)
                .setPositiveButton(
                    android.R.string.ok,
                    (dialog, which) -> {
                      taskAttachmentDao.delete(taskAttachment);
                      FileHelper.delete(context, taskAttachment.parseUri());
                      attachmentContainer.removeView(fileRow);
                    })
                .setNegativeButton(android.R.string.cancel, null)
                .show());
  }

  @Override
  protected void inject(FragmentComponent component) {
    component.inject(this);
  }

  @SuppressLint("NewApi")
  private void showFile(final TaskAttachment m) {
    FileHelper.startActionView(getActivity(), m.parseUri());
  }

  private void copyToAttachmentDirectory(Uri input) {
    Uri output = copyToUri(context, preferences.getAttachmentsDirectory(), input);
    TaskAttachment attachment =
        createNewAttachment(taskUuid, output, FileHelper.getFilename(context, output));
    taskAttachmentDao.createNew(attachment);
    addAttachment(attachment);
  }
}

File: app/src/main/java/com/todoroo/astrid/files/FilesControlSet.kt
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.files

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.tasks.R
import org.tasks.Strings
import org.tasks.compose.edit.AttachmentRow
import org.tasks.data.dao.TaskAttachmentDao
import org.tasks.data.entity.TaskAttachment
import org.tasks.dialogs.AddAttachmentDialog
import org.tasks.extensions.Context.takePersistableUriPermission
import org.tasks.files.FileHelper
import org.tasks.preferences.Preferences
import org.tasks.ui.TaskEditControlFragment
import timber.log.Timber
import javax.inject.Inject

@AndroidEntryPoint
class FilesControlSet : TaskEditControlFragment() {
    @Inject lateinit var taskAttachmentDao: TaskAttachmentDao
    @Inject lateinit var preferences: Preferences
    
    override fun createView(savedInstanceState: Bundle?) {
        val task = viewModel.viewState.value.task
        if (savedInstanceState == null) {
            if (task.hasTransitory(TaskAttachment.KEY)) {
                for (uri in (task.getTransitory<ArrayList<Uri>>(TaskAttachment.KEY))!!) {
                    newAttachment(uri)
                }
            }
        }
    }

    override fun bind(parent: ViewGroup?): View =
        (parent as ComposeView).apply {
            setContent {
                val viewState = viewModel.viewState.collectAsStateWithLifecycle().value
                AttachmentRow(
                    attachments = viewState.attachments,
                    openAttachment = {
                        Timber.d("Clicked open $it")
                        FileHelper.startActionView(
                            context,
                            if (Strings.isNullOrEmpty(it.uri)) null else Uri.parse(it.uri)
                        )
                    },
                    deleteAttachment = {
                        Timber.d("Clicked delete $it")
                        viewModel.setAttachments(viewState.attachments - it)
                    },
                    addAttachment = {
                        Timber.d("Add attachment clicked")
                        AddAttachmentDialog.newAddAttachmentDialog(this@FilesControlSet)
                            .show(parentFragmentManager, FRAG_TAG_ADD_ATTACHMENT_DIALOG)
                    },
                )
            }
        }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == AddAttachmentDialog.REQUEST_CAMERA || requestCode == AddAttachmentDialog.REQUEST_AUDIO) {
            if (resultCode == Activity.RESULT_OK) {
                lifecycleScope.launch {
                    val uri = data!!.data
                    copyToAttachmentDirectory(uri)
                    FileHelper.delete(requireContext(), uri)
                }
            }
        } else if (requestCode == AddAttachmentDialog.REQUEST_STORAGE || requestCode == AddAttachmentDialog.REQUEST_GALLERY) {
            if (resultCode == Activity.RESULT_OK) {
                lifecycleScope.launch {
                    val clip = data!!.clipData
                    if (clip != null) {
                        for (i in 0 until clip.itemCount) {
                            val item = clip.getItemAt(i)
                            requireContext().takePersistableUriPermission(item.uri)
                            copyToAttachmentDirectory(item.uri)
                        }
                    } else {
                        requireContext().takePersistableUriPermission(data.data!!)
                        copyToAttachmentDirectory(data.data)
                    }
                }
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    private suspend fun copyToAttachmentDirectory(input: Uri?) {
        val destination = preferences.attachmentsDirectory ?: return
        newAttachment(
            if (FileHelper.isInTree(requireContext(), destination, input!!)) {
                Timber.d("$input already exists in $destination")
                input
            } else {
                Timber.d("Copying $input to $destination")
                FileHelper.copyToUri(requireContext(), destination, input)
            }
        )
    }

    private fun newAttachment(output: Uri) {
        val attachment = TaskAttachment(
            uri = output.toString(),
            name = FileHelper.getFilename(requireContext(), output)!!,
        )
        lifecycleScope.launch {
            taskAttachmentDao.insert(attachment)
            viewModel.setAttachments(
                viewModel.viewState.value.attachments +
                        (taskAttachmentDao.getAttachment(attachment.remoteId) ?: return@launch))
        }
    }

    companion object {
        val TAG = R.string.TEA_ctrl_files_pref
        private const val FRAG_TAG_ADD_ATTACHMENT_DIALOG = "frag_tag_add_attachment_dialog"
    }
}
